Extends Piano Genie API with various conditioning signals

chrisdonahue opened the pull request:
Refactors Piano Genie API creating new subclasses `PianoGenieChord`, `PianoGenieKeysig`, `PianoGenieChordKeysig` which add various types of musical context as conditioning signals for the RNN. The original API for class `PianoGenie` is preserved without modification in this refactor.

chrisdonahue commented:
@notwaldorf I've kept `PitchClass` instead of `NoteSequence.PitchName` for now because I couldn't find any easy way to convert the latter enum into a number 1-12 indicating (indicating pitch classes C-B).

Also, I will update the demo with chord conditioning once all of the models are finished training. I would appreciate a review on these standalone API changes for now as my contract is expiring soon :)

notwaldorf commented:
Reviewing right now!

notwaldorf commented on music/src/piano_genie/model.ts:
```diff
@@ -291,6 +361,75 @@ class PianoGenie {
+  private time: Date;
+  private deltaTimeOverride: number;
+
+  // Inherits parent docstring.
```
Yeah I don't think any of these "Inherits" comments are needed

chrisdonahue replied:
Wasn't sure about this. Thanks!

notwaldorf commented on music/src/piano_genie/model.ts:
```diff
@@ -291,6 +361,75 @@ class PianoGenie {
+/**
+ * Original Piano Genie config with autoregression and delta time features.
+ */
+class PianoGenieAutoDt extends PianoGenieBase {
```
Since this is getting aliased anyway, can it maybe use the full name to be more clear? `PianoGenieAutoregressionDeltaTime`?

chrisdonahue replied:
Yes!

chrisdonahue commented:
@notwaldorf I've addressed your comments. I appreciate the quick review!

notwaldorf commented:
Do you need me to merge and cut a release?

chrisdonahue commented:
A new release is not necessary. Merging is optional. I can also just update this PR with the demo code when it is ready. Up to you.

notwaldorf commented:
Without a release you can't really use it, but that sounds ok with me. Just let me know when you want me to look again

chrisdonahue commented:
Hi Monica. I added an example demo using the new conditioned models and API. The model switching and conditioning controls complicated the demo substantially. As such, I made it a separate demo called `piano_genie_conditioned` and preserved the original (simpler) `piano_genie` demo. Ready for your eyes when you have a minute :)

chrisdonahue commented:
The new demo can be previewed at: https://chrisdonahue.com/magentajs_private/music/demos/piano_genie_conditioned.html

notwaldorf commented on music/demos/index.html:
```diff
@@ -43,6 +43,7 @@ <h3>
         <li><a href="groovae.html">Humanizing drum performances with GrooVAE</a></li>
         <li><a href="transcription.html">Piano transcription with Onsets and Frames</a></li>
         <li><a href="piano_genie.html">Enable novice musical improvisation with Piano Genie</a></li>
+        <li><a href="piano_genie_conditioned.html">Play Piano Genie with specified key sigantures and/or chord progressions</a></li>
```
typo in "signatures"

notwaldorf commented on music/demos/piano_genie_conditioned.ts:
```diff
@@ -0,0 +1,244 @@
+const GENIE_DIR = `${CHECKPOINTS_DIR}/piano_genie/model`;
+const GENIE_CHECKPOINTS: {[key:string]: {[key:string]: string}} = {
+  // Models conditioned on performance timing features
+  dt_only: {
```
please use less abbreviations and more words on these names, since they're getting shown in the select box, and not everybody knows what they stand for

notwaldorf commented on music/demos/piano_genie_conditioned.ts:
```diff
@@ -0,0 +1,244 @@
+  }
+  genie = undefined;
+
+  if (condtype === 'dt_only') {
```
you can have a `switch()` statement here

notwaldorf commented on music/demos/piano_genie_conditioned.ts:
```diff
@@ -0,0 +1,244 @@
+
+function initModelSelector() {
+  const modelSelector = document.getElementById('models') as HTMLSelectElement;
+  for (const c in GENIE_CHECKPOINTS) {
```
maybe call these checkpoint and index instead of c and n?

notwaldorf commented on music/demos/piano_genie_conditioned.ts:
```diff
@@ -0,0 +1,244 @@
+ * limitations under the License.
+ */
+
+import * as Tone from 'tone';
```
tone is exported as `mm.Player.tone` btw

chrisdonahue commented:
Thanks for pointing out my lack of any documentation about the various checkpoints. I have filtered the checkpoint list down to recommend ones for each conditioning configuration, and added a verbose comment about the other checkpoints for devs who want to venture off the beaten path.

Updated live preview: https://chrisdonahue.com/magentajs_private/music/demos/piano_genie_conditioned.html
